calyx/src/passes/compile_ref.rs

```diff
@@ -13,8 +13,34 @@ use std::rc::Rc;
 /// 2. Inline all the ports of the ref cells to the component signature
 /// 3. Remove all the ref cell mappings from the invoke statement
 /// 4. Inline all the mappings of ports to the invoke signature
+
+/// Map for storing added ports for each ref cell
+/// level of Hashmap represents:
+/// HashMap<-component name-, Hashmap<-cell name-, HashMap<-port name-, port>>>;
+type RefPortMap =
+    HashMap<ir::Id, HashMap<ir::Id, HashMap<ir::Id, RRC<ir::Port>>>>;
```

Comment:
Use `port.canonical` here to denest the HashMap by one level

Reply:
Seems like to use Canonical as a key of a hash map we would have to implement Compare, or Equals?

Reply:
I think that should be fine to `derive` on `ir::Canonical`